### PR TITLE
tls: introduce primitive tls and x509 sets

### PR DESCRIPTION
**Pull request by amery**
* `tls/x509utils/certpool.CertSet`
* `tls/store.CertSet`
* `tls/store.KeySet`

## Summary by CodeRabbit

- **New Features**
	- Introduced `CertSet` and `KeySet` types for managing collections of certificates and private keys, respectively.
	- Added functions for creating, initializing, and managing these sets, enhancing key and certificate management capabilities.
	- Implemented helper functions for hashing and matching certificates and keys.
	- Added a new package for a generic programmable TLS store.

- **Chores**
	- Updated dependencies to reflect new versions, ensuring compatibility and improved functionality.

### CODE REVIEW

File: tls/store/basic/cert_set.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	case cert == nil, cert.Leaf == nil:
+		return nil, core.ErrInvalid
+	default:
+		return cert.Leaf, nil
+	}
+}
+
+func certSetItemMatch(leaf *x509.Certificate, cert *tls.Certificate) bool {
+	return cert.Leaf.Equal(leaf)
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/store/basic/cert_set.go:72]: certSetItemMatch could also be called with pre-validated arguments

File: tls/store/basic/key_set.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	"darvaza.org/core"
+	"darvaza.org/x/container/set"
+
+	"darvaza.org/x/tls/x509utils"
+	"darvaza.org/x/tls/x509utils/certpool"
+)
+
+// KeySet keeps a thread-safe set of unique [x509utils.PrivateKey]s.
+type KeySet struct {
+	set.Set[x509utils.PublicKey, certpool.Hash, x509utils.PrivateKey]
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/store/basic/key_set.go:13]: The Set is supposed to be thread-safe, the List isn't

**Comment by amery** [comment]
@coderabbitai full review

**Comment by amery** [comment]
fixed typo on Clone()'s description

**Review by karasz — APPROVED** [review]
LGTM

**Comment by amery** [comment]
`darvaza.org/x/tls@v0.3.1` tagged

**Pull request merged**